examples/affinity.cpp

```diff
@@ -0,0 +1,225 @@
+
+    // 2. calculate the variance used in similarity function
+    size_t totDataPoints = dataCountAgg.get_value();
+    ndvec dataMean = dataxAgg.get_value() / totDataPoints;
+    ndvec dataVar = dataxxAgg.get_value() / totDataPoints;
+    // Var(x) = <x^2> - <x><x>
+    for (int i = 0; i < dimension; ++i) {
+        dataVar[i] -= dataMean[i] * dataMean[i];
+        husky::base::log_msg("var " + std::to_string(i) + " " + std::to_string(dataVar[i]));
+    }
+
+    // 3. calculate the affinity matrix
+    husky::ObjList<MatElem> affMatElem_list;
+    globalize(affMatElem_list);
+    auto& pairCh = husky::ChannelFactory::create_push_channel<NDpoint>(ndpoint_list, affMatElem_list);
+
+    // push 2 data points to each matrix element above the diagonal
+    list_execute(ndpoint_list, [&](NDpoint& p) {
+        // the pid-th col
+        for (uint i = 0; i < p.id(); ++i) {
+            pairCh.push(p, i * totDataPoints + p.id());
+        }
+        // the pid-th row
+        for (uint i = p.id() + 1; i < totDataPoints; ++i) {
+            pairCh.push(p, p.id() * totDataPoints + i);
+        }
+    });
+
+    husky::base::log_msg("affMat size " + std::to_string(affMatElem_list.get_size()));
+
+    husky::ObjList<MatRow> affMatRow_list;
+    auto& elemToRowCh = husky::ChannelFactory::create_push_channel<MatElem>(affMatElem_list, affMatRow_list);
+
+    // group matrix elements to rows
+    list_execute(affMatElem_list, [&](MatElem& e) {
+        auto pts = pairCh.get(e);
+        float nSigmaSq = 0;
+        for (uint i = 0; i < dimension; i++) {
+            nSigmaSq += pow(pts[0].coords[i] - pts[1].coords[i], 2) / (2. * dataVar[i]);
+        }
+        e.val = exp(-nSigmaSq);
+        int col = e.id() % totDataPoints;
+        int row = e.id() / totDataPoints;
+        // As affinity matrix is symmetric, each matrix element above the diagonal "belongs" to 2 rows
+        elemToRowCh.push(e, col);
+        elemToRowCh.push(e, row);
+    });
+
+    // 4. print output
+    list_execute(affMatRow_list, [&](MatRow& r) {
+        auto elems = elemToRowCh.get(r);
+        std::vector<float> rowVec(totDataPoints);
+        for (auto aElem : elems) {
+            int elemCol = aElem.id() % totDataPoints;
+            int elemRow = aElem.id() / totDataPoints;
+            if (elemCol > r.id())
+                rowVec[elemCol] = aElem.val;
+            else
+                rowVec[elemRow] = aElem.val;
+        }
+        std::string log;
+        log = std::to_string(r.id());
+        for (auto aElem : rowVec) {
+            log += "  " + std::to_string(aElem);
+        }
+        log += "\n";
+        husky::io::HDFS::Write("master", "9000", log, husky::Context::get_param("outDir"),
+                               husky::Context::get_global_tid());
+    });
+}
+
+int main(int argc, char** argv) {
+    std::vector<std::string> args;
+    args.push_back("hdfs_namenode");
+    args.push_back("hdfs_namenode_port");
+    args.push_back("input");
```

Comment:
Can you indicate the format of the input and output data here? For example, what input data the user should prepare and what output data this program should generate?